File: frontend/src/api/knowledge.ts
import { apiClient, unwrapData } from './client';
import { APIResponse, KnowledgeDocument } from '@/types/api';

export interface UploadDocumentPayload {
  title: string;
  doc_type: string; // runbook | architecture | postmortem | playbooks | docs
  content: string;
  category?: string;
  tags?: string[];
}

export interface RAGSearchPayload {
  query: string;
  service_name?: string;
  severity?: string;
  top_k?: number;
}

export const knowledgeApi = {
  async uploadDocument(projectId: string, payload: UploadDocumentPayload): Promise<KnowledgeDocument> {
    const res = await apiClient.post<APIResponse<KnowledgeDocument>>('/knowledge/upload', payload, {
      params: { project_id: projectId },
    });
    return unwrapData(res);
  },

  async listDocuments(projectId: string, docType?: string): Promise<KnowledgeDocument[]> {
    const res = await apiClient.get<APIResponse<KnowledgeDocument[]>>('/knowledge/documents', {
      params: { project_id: projectId, doc_type: docType },
    });
    return unwrapData(res);
  },

  async searchRAG(projectId: string, payload: RAGSearchPayload): Promise<any> {
    const res = await apiClient.post<APIResponse<any>>('/knowledge/search', payload, {
      params: { project_id: projectId },
    });
    return unwrapData(res);
  },
};

File: backend/api/v1/knowledge.py
import uuid
from typing import List, Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.api.dependencies import get_current_user
from backend.database.session import get_async_session
from backend.models.users import User
from backend.schemas.common import APIResponse
from backend.schemas.knowledge import KnowledgeDocumentCreate, KnowledgeDocumentResponse, RAGQuerySchema
from backend.services.knowledge_service import knowledge_service

router = APIRouter(prefix="/knowledge", tags=["Knowledge Base & Enterprise RAG"])


@router.post("/upload", response_model=APIResponse[KnowledgeDocumentResponse])
async def upload_document(
    project_id: uuid.UUID,
    data: KnowledgeDocumentCreate,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Uploads runbooks, architecture docs, or playbooks. Automatically chunks, embeds, and indexes into ChromaDB."""
    doc = await knowledge_service.ingest_document(session, project_id, current_user.organization_id, data)
    return APIResponse(
        message="Document uploaded, chunked, embedded, and indexed into ChromaDB vectorstore.",
        data=KnowledgeDocumentResponse.model_validate(doc)
    )


@router.get("/documents", response_model=APIResponse[List[KnowledgeDocumentResponse]])
async def list_documents(
    project_id: uuid.UUID = Query(...),
    doc_type: Optional[str] = Query(None),
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Lists indexed knowledge documents for a project."""
    docs = await knowledge_service.list_documents(session, project_id, current_user.organization_id, doc_type=doc_type)
    return APIResponse(
        message="Indexed documents list retrieved.",
        data=[KnowledgeDocumentResponse.model_validate(d) for d in docs]
    )


@router.post("/search", response_model=APIResponse[dict])
async def search_rag_knowledge(
    project_id: uuid.UUID,
    data: RAGQuerySchema,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Executes RAG hybrid similarity query over ChromaDB knowledge collections."""
    result = await knowledge_service.search_rag_knowledge(
        session=session,
        project_id=project_id,
        organization_id=current_user.organization_id,
        query=data.query,
        service_name=data.service_name,
        severity=data.severity,
        top_k=data.top_k
    )
    return APIResponse(
        message="RAG search completed.",
        data=result
    )
